File: src/syntax_cache.zig
const std = @import("std");
const diff = @import("diff.zig");
const git = @import("git.zig");
const syntax = @import("syntax.zig");
const syntax_grammars = @import("syntax_grammars.zig");
const syntax_query = @import("syntax_query.zig");
const theme = @import("theme.zig");
const tree_sitter = @import("tree_sitter.zig");
const util = @import("util.zig");

pub const Error = error{
    SyntaxUnavailable,
    SourceTooLarge,
} || std.mem.Allocator.Error || git.GitError || tree_sitter.Error;

const max_file_size = 512 * 1024;

pub const SyntaxCache = struct {
    allocator: std.mem.Allocator,
    io: std.Io,
    repo: diff.Repository,
    target: diff.ReviewTarget,
    debug_git: bool,
    entries: std.ArrayList(Entry) = .empty,

    const Status = enum { ready, unavailable, too_large };

    const Entry = struct {
        file_index: usize,
        side: git.FileSide,
        status: Status,
        highlight: ?syntax_query.SideHighlight,

        fn deinit(self: *Entry, allocator: std.mem.Allocator) void {
            if (self.highlight) |*highlight| highlight.deinit(allocator);
        }
    };

    const Outcome = union(enum) {
        ready: syntax_query.SideHighlight,
        unavailable,
        too_large,
    };

    pub fn init(
        allocator: std.mem.Allocator,
        io: std.Io,
        repo: diff.Repository,
        target: diff.ReviewTarget,
        debug_git: bool,
    ) SyntaxCache {
        return .{
            .allocator = allocator,
            .io = io,
            .repo = repo,
            .target = target,
            .debug_git = debug_git,
        };
    }

    pub fn deinit(self: *SyntaxCache) void {
        for (self.entries.items) |*entry| entry.deinit(self.allocator);
        self.entries.deinit(self.allocator);
    }

    pub fn highlightDiffLine(
        self: *SyntaxCache,
        allocator: std.mem.Allocator,
        ansi: theme.Ansi,
        palette: theme.ThemeTokens,
        file_index: usize,
        file: diff.DiffFile,
        line: *const diff.DiffLine,
    ) Error![]u8 {
        if (!ansi.enabled) return error.SyntaxUnavailable;
        const line_ref = lineRef(line) orelse return error.SyntaxUnavailable;
        const highlight = try self.get(file_index, file, line_ref.side);
        const source_text = highlight.lineText(line_ref.line_number) orelse return error.SyntaxUnavailable;
        if (!util.eql(source_text, line.text)) return error.SyntaxUnavailable;
        return syntax.renderHighlightedLine(allocator, ansi, palette, line.text, highlight.lineSpans(line_ref.line_number));
    }

    fn get(self: *SyntaxCache, file_index: usize, file: diff.DiffFile, side: git.FileSide) Error!*syntax_query.SideHighlight {
        for (self.entries.items) |*entry| {
            if (entry.file_index == file_index and entry.side == side) {
                return switch (entry.status) {
                    .ready => &entry.highlight.?,
                    .unavailable => error.SyntaxUnavailable,
                    .too_large => error.SourceTooLarge,
                };
            }
        }
        // Compute the outcome once and cache it — including the negative results — so an
        // unavailable or too-large side never re-spawns git / re-reads the blob on later frames.
        var outcome = try self.computeOutcome(file, side);
        switch (outcome) {
            .ready => {
                errdefer outcome.ready.deinit(self.allocator);
                try self.entries.append(self.allocator, .{
                    .file_index = file_index,
                    .side = side,
                    .status = .ready,
                    .highlight = outcome.ready,
                });
                return &self.entries.items[self.entries.items.len - 1].highlight.?;
            },
            .unavailable => {
                try self.entries.append(self.allocator, .{ .file_index = file_index, .side = side, .status = .unavailable, .highlight = null });
                return error.SyntaxUnavailable;
            },
            .too_large => {
                try self.entries.append(self.allocator, .{ .file_index = file_index, .side = side, .status = .too_large, .highlight = null });
                return error.SourceTooLarge;
            },
        }
    }

    fn computeOutcome(self: *SyntaxCache, file: diff.DiffFile, side: git.FileSide) Error!Outcome {
        const language = file.language orelse return .unavailable;
        const grammar = syntax_grammars.find(language) orelse return .unavailable;
        const source = try git.loadFileSide(self.allocator, self.io, self.repo, self.target, file, side, self.debug_git) orelse return .unavailable;
        if (source.len > max_file_size) {
            self.allocator.free(source);
            return .too_large;
        }
        // build() takes ownership of source and frees it on its own error paths.
        const highlight = try syntax_query.build(self.allocator, grammar.language(), grammar.query, source);
        return .{ .ready = highlight };
    }
};

const LineRef = struct {
    side: git.FileSide,
    line_number: u32,
};

fn lineRef(line: *const diff.DiffLine) ?LineRef {
    return switch (line.kind) {
        .delete => if (line.old_lineno) |n| .{ .side = .old, .line_number = n } else null,
        .add => if (line.new_lineno) |n| .{ .side = .new, .line_number = n } else null,
        .context => if (line.new_lineno) |n|
            .{ .side = .new, .line_number = n }
        else if (line.old_lineno) |n|
            .{ .side = .old, .line_number = n }
        else
            null,
        .meta => null,
    };
}

test "line refs map diff sides" {
    var empty: [0]u8 = .{};
    var add = diff.DiffLine{ .kind = .add, .old_lineno = null, .new_lineno = 12, .text = empty[0..], .stable_line_id = empty[0..] };
    var del = diff.DiffLine{ .kind = .delete, .old_lineno = 7, .new_lineno = null, .text = empty[0..], .stable_line_id = empty[0..] };
    try std.testing.expectEqual(git.FileSide.new, lineRef(&add).?.side);
    try std.testing.expectEqual(@as(u32, 12), lineRef(&add).?.line_number);
    try std.testing.expectEqual(git.FileSide.old, lineRef(&del).?.side);
}

File: src/syntax_cache.rs
use std::collections::HashMap;

use crate::{
    Error, Result,
    diff::{DiffFile, DiffLine, DiffLineKind, Repository, ReviewTarget},
    git::{self, FileSide},
    syntax, syntax_grammars,
    syntax_query::{self, SideHighlight},
    theme::{Ansi, ThemeTokens},
};

#[derive(Clone, Debug)]
pub struct SyntaxCache {
    repo: Repository,
    target: ReviewTarget,
    debug_git: bool,
    entries: HashMap<(usize, FileSide), Outcome>,
}

#[derive(Clone, Debug)]
enum Outcome {
    Ready(SideHighlight),
    Unavailable,
    TooLarge,
}

impl SyntaxCache {
    pub fn new(repo: &Repository, target: &ReviewTarget, debug_git: bool) -> Self {
        Self {
            repo: repo.clone(),
            target: target.clone(),
            debug_git,
            entries: HashMap::new(),
        }
    }

    pub fn highlight_diff_line(
        &mut self,
        ansi: Ansi,
        palette: ThemeTokens,
        file_index: usize,
        file: &DiffFile,
        line: &DiffLine,
    ) -> Result<String> {
        if !ansi.enabled {
            return Err(Error::SyntaxUnavailable);
        }
        let (side, number) = line_ref(line).ok_or(Error::SyntaxUnavailable)?;
        let highlight = self.get(file_index, file, side)?;
        if highlight.line_text(number) != Some(line.text.as_bytes()) {
            return Err(Error::SyntaxUnavailable);
        }
        Ok(syntax::render_highlighted_line(
            ansi,
            palette,
            &line.text,
            highlight.line_spans(number),
        ))
    }

    fn get(
        &mut self,
        file_index: usize,
        file: &DiffFile,
        side: FileSide,
    ) -> Result<&SideHighlight> {
        let key = (file_index, side);
        // A cache belongs to one immutable snapshot. Cache negative outcomes as
        // well, so repainting a missing/oversized side never repeats Git or I/O.
        if !self.entries.contains_key(&key) {
            let outcome = self.compute_outcome(file, side)?;
            self.entries.insert(key, outcome);
        }
        match &self.entries[&key] {
            Outcome::Ready(highlight) => Ok(highlight),
            Outcome::Unavailable => Err(Error::SyntaxUnavailable),
            Outcome::TooLarge => Err(Error::SourceTooLarge),
        }
    }

    fn compute_outcome(&self, file: &DiffFile, side: FileSide) -> Result<Outcome> {
        let Some(grammar) = file.language.as_deref().and_then(syntax_grammars::find) else {
            return Ok(Outcome::Unavailable);
        };
        let source = match git::load_file_side(&self.repo, &self.target, file, side, self.debug_git)
        {
            Ok(Some(source)) => source,
            Ok(None) | Err(Error::SyntaxUnavailable) => return Ok(Outcome::Unavailable),
            Err(Error::SourceTooLarge) => return Ok(Outcome::TooLarge),
            Err(err) => return Err(err),
        };
        match syntax_query::build(grammar.language(), grammar.query, source) {
            Ok(highlight) => Ok(Outcome::Ready(highlight)),
            Err(Error::SourceTooLarge) => Ok(Outcome::TooLarge),
            Err(Error::SyntaxUnavailable) => Ok(Outcome::Unavailable),
            Err(err) => Err(err),
        }
    }
}

fn line_ref(line: &DiffLine) -> Option<(FileSide, u32)> {
    match line.kind {
        DiffLineKind::Delete => line.old_lineno.map(|number| (FileSide::Old, number)),
        DiffLineKind::Add => line.new_lineno.map(|number| (FileSide::New, number)),
        DiffLineKind::Context => line
            .new_lineno
            .map(|number| (FileSide::New, number))
            .or_else(|| line.old_lineno.map(|number| (FileSide::Old, number))),
        DiffLineKind::Meta => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        diff::{self, DiffSource, FileStatus},
        theme::catppuccin_mocha,
    };
    use std::{fs, process::Command};

    const ANSI: Ansi = Ansi {
        enabled: true,
        true_color: true,
    };

    fn line(kind: DiffLineKind, old: Option<u32>, new: Option<u32>, text: &str) -> DiffLine {
        DiffLine {
            kind,
            old_lineno: old,
            new_lineno: new,
            text: text.into(),
            stable_line_id: String::new(),
        }
    }

    fn fixture() -> (tempfile::TempDir, Repository, ReviewTarget, DiffFile) {
        let directory = tempfile::tempdir().unwrap();
        let repo = Repository {
            root_path: directory.path().to_path_buf(),
            repo_id: "syntax-test".into(),
            current_branch: "test".into(),
        };
        let target = diff::make_review_target(&[]);
        let file = DiffFile {
            path: "sample.rs".into(),
            old_path: None,
            status: FileStatus::Added,
            source: DiffSource::Untracked,
            language: Some("rust".into()),
            is_binary: false,
            hunks: Vec::new(),
            patch_fingerprint: String::new(),
            patch_text: Vec::new(),
        };
        (directory, repo, target, file)
    }

    #[test]
    fn line_references_select_correct_sides() {
        assert_eq!(
            line_ref(&line(DiffLineKind::Add, None, Some(12), "")),
            Some((FileSide::New, 12))
        );
        assert_eq!(
            line_ref(&line(DiffLineKind::Delete, Some(7), None, "")),
            Some((FileSide::Old, 7))
        );
        assert_eq!(
            line_ref(&line(DiffLineKind::Context, Some(7), Some(12), "")),
            Some((FileSide::New, 12))
        );
        assert_eq!(
            line_ref(&line(DiffLineKind::Context, Some(7), None, "")),
            Some((FileSide::Old, 7))
        );
        assert_eq!(
            line_ref(&line(DiffLineKind::Meta, Some(7), Some(12), "")),
            None
        );
        assert_eq!(line_ref(&line(DiffLineKind::Add, None, None, "")), None);
    }

    #[test]
    fn successful_highlights_are_cached_and_mismatches_are_rejected() {
        let (_directory, repo, target, file) = fixture();
        let path = repo.root_path.join(&file.path);
        fs::write(&path, "fn main() {}\n").unwrap();
        let mut cache = SyntaxCache::new(&repo, &target, false);
        let original = line(DiffLineKind::Add, None, Some(1), "fn main() {}");
        let first = cache
            .highlight_diff_line(ANSI, catppuccin_mocha(), 0, &file, &original)
            .unwrap();
        assert!(first.contains("\x1b[38;2;"));
        fs::remove_file(&path).unwrap();
        assert_eq!(
            cache
                .highlight_diff_line(ANSI, catppuccin_mocha(), 0, &file, &original)
                .unwrap(),
            first
        );
        assert_eq!(cache.entries.len(), 1);
        let changed = line(DiffLineKind::Add, None, Some(1), "fn changed() {}");
        assert!(matches!(
            cache.highlight_diff_line(ANSI, catppuccin_mocha(), 0, &file, &changed),
            Err(Error::SyntaxUnavailable)
        ));
        let absent = line(DiffLineKind::Add, None, Some(2), "fn main() {}");
        assert!(matches!(
            cache.highlight_diff_line(ANSI, catppuccin_mocha(), 0, &file, &absent),
            Err(Error::SyntaxUnavailable)
        ));
    }

    #[test]
    fn missing_and_too_large_sources_are_negative_cached() {
        let (_directory, repo, target, file) = fixture();
        let path = repo.root_path.join(&file.path);
        let mut cache = SyntaxCache::new(&repo, &target, false);
        assert!(matches!(
            cache.get(0, &file, FileSide::New),
            Err(Error::SyntaxUnavailable)
        ));
        fs::write(&path, "fn main() {}\n").unwrap();
        assert!(matches!(
            cache.get(0, &file, FileSide::New),
            Err(Error::SyntaxUnavailable)
        ));
        assert!(cache.get(1, &file, FileSide::New).is_ok());
        fs::write(&path, vec![b' '; syntax_query::MAX_FILE_SIZE + 1]).unwrap();
        assert!(matches!(
            cache.get(2, &file, FileSide::New),
            Err(Error::SourceTooLarge)
        ));
        fs::write(&path, "fn main() {}\n").unwrap();
        assert!(matches!(
            cache.get(2, &file, FileSide::New),
            Err(Error::SourceTooLarge)
        ));
        assert_eq!(cache.entries.len(), 3);
    }

    #[test]
    fn unsupported_languages_and_disabled_ansi_do_not_load_source() {
        let (_directory, repo, target, mut file) = fixture();
        let mut cache = SyntaxCache::new(&repo, &target, false);
        let added = line(DiffLineKind::Add, None, Some(1), "fn main() {}");
        assert!(matches!(
            cache.highlight_diff_line(
                Ansi {
                    enabled: false,
                    ..ANSI
                },
                catppuccin_mocha(),
                0,
                &file,
                &added
            ),
            Err(Error::SyntaxUnavailable)
        ));
        assert!(cache.entries.is_empty());
        file.language = Some("markdown".into());
        assert!(matches!(
            cache.get(0, &file, FileSide::New),
            Err(Error::SyntaxUnavailable)
        ));
        assert!(matches!(
            cache.entries[&(0, FileSide::New)],
            Outcome::Unavailable
        ));
    }

    #[test]
    fn old_index_and_new_worktree_have_independent_caches() {
        let (_directory, repo, target, mut file) = fixture();
        for args in [
            &["init", "-q"][..],
            &["config", "core.autocrlf", "false"][..],
        ] {
            let result = Command::new("git")
                .args(args)
                .current_dir(&repo.root_path)
                .output()
                .unwrap();
            assert!(
                result.status.success(),
                "{}",
                String::from_utf8_lossy(&result.stderr)
            );
        }
        let path = repo.root_path.join(&file.path);
        fs::write(&path, "fn old() {}\n").unwrap();
        let result = Command::new("git")
            .args(["add", "--", &file.path])
            .current_dir(&repo.root_path)
            .output()
            .unwrap();
        assert!(
            result.status.success(),
            "{}",
            String::from_utf8_lossy(&result.stderr)
        );
        fs::write(&path, "fn new() {}\n").unwrap();
        file.source = DiffSource::Unstaged;
        file.status = FileStatus::Modified;
        let mut cache = SyntaxCache::new(&repo, &target, false);
        for (kind, old, new, text) in [
            (DiffLineKind::Delete, Some(1), None, "fn old() {}"),
            (DiffLineKind::Add, None, Some(1), "fn new() {}"),
        ] {
            let rendered = cache
                .highlight_diff_line(
                    ANSI,
                    catppuccin_mocha(),
                    0,
                    &file,
                    &line(kind, old, new, text),
                )
                .unwrap();
            assert!(rendered.contains("\x1b[38;2;"));
        }
        assert_eq!(cache.entries.len(), 2);
    }
}
